drivers/crypto/crypto_it8xxx2_sha_v2.c

```diff
@@ -0,0 +1,280 @@
+
+#define IT8XXX2_SHA_REGS_BASE DT_REG_ADDR(DT_NODELABEL(sha0))
+
+/* 0x00: SHA Control Register */
+#define IT8XXX2_REG_SHACR        (0x00)
+#define IT8XXX2_SEL1SHA1         BIT(6)
+#define IT8XXX2_SELSHA2ALL       (BIT(5) | BIT(4))
+#define IT8XXX2_SHAWB            BIT(2)
+#define IT8XXX2_SHAINI           BIT(1)
+#define IT8XXX2_SHAEXE           BIT(0)
+/* 0x01: SHA Status Register */
+#define IT8XXX2_REG_SHASR        (0x01)
+#define IT8XXX2_SHAIE            BIT(3)
+#define IT8XXX2_SHAIS            BIT(2)
+#define IT8XXX2_SHABUSY          BIT(0)
+/* 0x02: SHA Execution Counter Register */
+#define IT8XXX2_REG_SHAECR       (0x02)
+#define IT8XXX2_SHAEXEC_64Byte   0x0
+#define IT8XXX2_SHAEXEC_512Byte  0x7
+#define IT8XXX2_SHAEXEC_1KByte   0xf
+/* 0x03: SHA DLM Base Address 0 Register */
+#define IT8XXX2_REG_SHADBA0R     (0x03)
+/* 0x04: SHA DLM Base Address 1 Register */
+#define IT8XXX2_REG_SHADBA1R     (0x04)
+
+#define SHA_SHA256_HASH_LEN        32
+#define SHA_SHA256_BLOCK_LEN       64
+#define SHA_SHA256_SRAM_BUF        1024
+#define SHA_SHA256_HASH_LEN_WORDS  (SHA_SHA256_HASH_LEN / sizeof(uint32_t))
+#define SHA_SHA256_BLOCK_LEN_WORDS (SHA_SHA256_BLOCK_LEN / sizeof(uint32_t))
+#define SHA_SHA256_SRAM_BUF_WORDS  (SHA_SHA256_SRAM_BUF / sizeof(uint32_t))
+
+/*
+ * This struct is used by the hardware and must be stored in RAM first 4k-byte
+ * and aligned on a 256-byte boundary.
+ */
+struct chip_sha256_ctx {
+	union {
+		/* SHA data buffer */
+		uint32_t w_sha[SHA_SHA256_SRAM_BUF_WORDS];
+		uint8_t w_input[SHA_SHA256_SRAM_BUF];
+	};
+	/* H[0] ~ H[7] */
+	uint32_t h[SHA_SHA256_HASH_LEN_WORDS];
+	uint32_t sha_init;
+	uint32_t w_input_index;
+	uint32_t total_len;
+} __aligned(256);
+
+Z_GENERIC_SECTION(.__sha256_ram_block) struct chip_sha256_ctx chip_ctx;
+
+static void it8xxx2_sha256_init(bool init_k)
+{
+	chip_ctx.sha_init = init_k;
+	chip_ctx.w_input_index = 0;
+	chip_ctx.total_len = 0;
+
+	/* Set DLM address for input data */
+	sys_write8(((uint32_t)&chip_ctx) & 0xc0,
+			IT8XXX2_SHA_REGS_BASE + IT8XXX2_REG_SHADBA0R);
+	sys_write8(((uint32_t)&chip_ctx) >> 8,
+			IT8XXX2_SHA_REGS_BASE + IT8XXX2_REG_SHADBA1R);
+}
+
+static void it8xxx2_sha256_module_calculation(void)
+{
+	struct gctrl_it8xxx2_regs *const gctrl_regs = GCTRL_IT8XXX2_REGS_BASE;
+	uint32_t key;
+	uint8_t sha_ctrl;
+
+	sha_ctrl = sys_read8(IT8XXX2_SHA_REGS_BASE + IT8XXX2_REG_SHACR);
+
+	if (chip_ctx.sha_init) {
+		sha_ctrl |= (IT8XXX2_SHAINI | IT8XXX2_SHAEXE);
+		chip_ctx.sha_init = 0;
+	} else {
+		sha_ctrl |= IT8XXX2_SHAEXE;
+	}
+
+	/*
+	 * Global interrupt is disabled because the CPU cannot access memory
+	 * via the DLM (Data Local Memory) bus while HW module is computing
+	 * hash.
+	 */
+	key = irq_lock();
+	/* Crypto use SRAM */
+	gctrl_regs->GCTRL_PMER3 |= IT8XXX2_GCTRL_SRAM_CRYPTO_USED;
+	sys_write8(sha_ctrl, IT8XXX2_SHA_REGS_BASE + IT8XXX2_REG_SHACR);
+
+	while (sys_read8(IT8XXX2_SHA_REGS_BASE + IT8XXX2_REG_SHASR) & IT8XXX2_SHABUSY)
+		;
```

Comment:
Unbounded loop.   This is especially dangerous because you've disabled interrupts.

Reply:
How long is the `SHABUSY` expected to be set?   Is there any recovery possible if the SHA module hangs?

Reply:
I confirmed with HW designer about the calculation time:
HW 64 bytes data calculation ~= 4us;
HW 1024 bytes data calculation ~= 66us.
HW write back the hash result to DLM ~= 1us.

So I change the while(); loop to finite for loop with a timeout message.